// @contentful/app-scripts/lib/activate/activate-bundle.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.activateBundle = void 0;
const ora_1 = __importDefault(require("ora"));
const chalk_1 = require("chalk");
const utils_1 = require("../utils");
const contentful_management_1 = require("contentful-management");
async function activateBundle({ accessToken, organization, definition, bundleId, host, hasFrontend, }) {
    const activationSpinner = (0, ora_1.default)('Activating your bundle').start();
    const plainClient = (0, contentful_management_1.createClient)({ accessToken, host }, { type: 'plain' });
    const defaultLocations = [{ location: 'dialog' }];
    let doesBundleHaveFrontend = hasFrontend;
    const currentDefinition = await plainClient.appDefinition.get({
        appDefinitionId: definition.value,
        organizationId: organization.value,
    });
    currentDefinition.bundle = {
        sys: {
            id: bundleId,
            linkType: 'AppBundle',
            type: 'Link',
        },
    };
    // if `hasFrontend` is not passed in, fetch the bundle to determine if it has a frontend
    let bundle;
    if (doesBundleHaveFrontend === undefined) {
        bundle = await plainClient.appBundle.get({
            appBundleId: bundleId,
            appDefinitionId: definition.value,
            organizationId: organization.value,
        });
        if (bundle.files.length) {
            doesBundleHaveFrontend = true;
        }
        else {
            doesBundleHaveFrontend = false;
        }
    }
    // if the bundle has frontend code in the bundle, remove the src from the definition since the frontend will now be hosted by Contentful
    if (doesBundleHaveFrontend) {
        delete currentDefinition.src;
    }
    // if the bundle has frontend code and there are no locations already on the app definition set the default location to dialog
    if (doesBundleHaveFrontend && !currentDefinition.locations?.length) {
        currentDefinition.locations = defaultLocations;
    }
    try {
        await plainClient.appDefinition.update({
            appDefinitionId: definition.value,
            organizationId: organization.value,
        }, currentDefinition);
    }
    catch (err) {
        let errorData;
        try {
            errorData = JSON.parse(err.message);
        }
        catch {
            errorData = null;
        }
        if (errorData?.status === 400 && errorData?.message?.includes('Function upload failed')) {
            // Extract first line of error message (the actual error type)
            const firstLine = errorData.message.split('\n')[0];
            const errorType = firstLine.replace('Function upload failed: ', '');
            // Clean display
            console.error(`\n${(0, chalk_1.bold)((0, chalk_1.yellow)('❌ Function Upload Failed'))}\n`);
            console.error(`${(0, chalk_1.bold)('Error:')} ${errorType}\n`);
            if (errorData.details?.errors?.[0]) {
                // Show just the first few lines of stack trace
                const firstError = errorData.details.errors[0];
                const lines = firstError.split('\n').slice(0, 3);
                console.error(lines.join('\n'));
            }
            console.error(`\n${(0, chalk_1.cyan)('💡 Tip:')} Line numbers may reference minified code, rebuild with ${(0, chalk_1.bold)('--no-minify')} to debug your source code:`);
            console.error(`   ${(0, chalk_1.bold)('npm run build:functions -- --no-minify && npm run upload')}\n`);
            // Don't re-throw (prevents Commander duplicate display)
            process.exit(1);
        }
        // Default error handling for other errors
        (0, utils_1.throwError)(err, 'Something went wrong activating your bundle. Make sure you used the correct definition-id.');
    }
    finally {
        activationSpinner.stop();
    }
    console.log(`
  ${(0, chalk_1.cyan)('Success!')} Your app bundle was activated for ${(0, chalk_1.cyan)(definition.name)} in ${(0, chalk_1.bold)(organization.name)}.
  Bundle Id: ${(0, chalk_1.yellow)(bundleId)}`);
    console.log(`
  ----------------------------
  Ready to share your app with the world? Submit it to the Developer Showcase (${(0, chalk_1.cyan)('https://ctfl.io/dev-showcase')}) or our Marketplace (${(0, chalk_1.cyan)('https://ctfl.io/submit-app')}).
 `);
}
exports.activateBundle = activateBundle;
